### PR TITLE
BUG: df.loc setitem-with-expansion with duplicate index

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

also CategoricalIndex.insert

### CODE REVIEW

**Comment by jbrockmendel** [comment]
cc @phofl

File: pandas/tests/indexing/test_loc.py [hunk 1]
```diff
@@ -1657,6 +1658,56 @@ def test_loc_setitem_with_expansion_inf_upcast_empty(self):
...
+        orig = DataFrame(arr, index=index, columns=[0])
+
+        # key that will requiring object-dtype casting in the index
+        key = "kapow"
+        assert key not in index  # otherwise test is invalid
+        # TODO: using a tuple key breaks here in many cases
+
+        exp_index = index.insert(len(index), key)
+        if isinstance(index, MultiIndex):
+            exp_index = index.insert(len(index), key)
```
Comment (phofl) [inline pandas/tests/indexing/test_loc.py:1683]: I think this is not needed?

LGTM otherwise

File: pandas/core/indexes/extension.py [hunk 2]
```diff
@@ -370,11 +374,16 @@ def insert(self: _T, loc: int, item) -> _T:
...
         """
         arr = self._data
-        code = arr._validate_scalar(item)
-
-        new_vals = np.concatenate((arr._ndarray[:loc], [code], arr._ndarray[loc:]))
-        new_arr = arr._from_backing_data(new_vals)
-        return type(self)._simple_new(new_arr, name=self.name)
+        try:
+            code = arr._validate_scalar(item)
+        except (ValueError, TypeError):
```
Comment (jreback) [inline pandas/core/indexes/extension.py:379]: can you add comments on these cases, e.g. what is hitting them / raising

File: pandas/tests/indexing/test_categorical.py [hunk 3]
```diff
@@ -64,11 +68,28 @@ def test_loc_scalar(self):
         df2.loc["d"] = 10
         tm.assert_frame_equal(df2, expected)
 
-        msg = "'fill_value=d' is not present in this Categorical's categories"
-        with pytest.raises(TypeError, match=msg):
-            df.loc["d", "A"] = 10
-        with pytest.raises(TypeError, match=msg):
-            df.loc["d", "C"] = 10
+        # Setting-with-expansion with a new key "d" that is not among caegories
```
Comment (jreback) [inline pandas/tests/indexing/test_categorical.py:71]: might be worth splitting this test (e.g. errors to a new tests) and even these cases

File: pandas/tests/indexing/test_loc.py [hunk 4]
```diff
@@ -1657,6 +1658,56 @@ def test_loc_setitem_with_expansion_inf_upcast_empty(self):
...
+            assert exp_index[-1][0] == key
+        else:
+            assert exp_index[-1] == key
+        exp_data = np.arange(N + 1).astype(np.float64)
+        expected = DataFrame(exp_data, index=exp_index, columns=[0])
+
+        # Add new row, but no new columns
+        df = orig.copy()
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py:1691]: same might be worth splitting this up a bit

File: pandas/core/indexing.py [hunk 5]
```diff
@@ -1641,7 +1641,12 @@ def _setitem_with_indexer(self, indexer, value, name="iloc"):
                     # so the object is the same
                     index = self.obj._get_axis(i)
                     labels = index.insert(len(index), key)
-                    self.obj._mgr = self.obj.reindex(labels, axis=i)._mgr
+                    taker = list(range(len(index))) + [-1]
```
Comment (jreback) [inline pandas/core/indexing.py:1644]: can you add comments on what is happening here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
I think we should deprecate indexing expansion, pretty sure we have an issue for this.

File: pandas/tests/indexing/test_loc.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_loc.py:1683]: yep, thanks

**Comment by jbrockmendel** [comment]
updted per comments + greenish

File: pandas/core/indexing.py (hunk 5 above)
Comment (jreback) [reply pandas/core/indexing.py:1648]: prob doesn't make much difference but can you just use np.arange here?

Comment (jbrockmendel) [reply pandas/core/indexing.py:1648]:
```
In [2]: %timeit list(np.arange(10))
1.6 µs ± 15.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit list(range(10))
238 ns ± 7.09 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Comment (jreback) [reply pandas/core/indexing.py:1648]: - sure that's small sample, is it always really small like this?
- needs to have ensure_int_platform generally (you can directly construct the np array like that)
- we *always* use np.arange for this so changing patterns is not helpful

Comment (jbrockmendel) [reply pandas/core/indexing.py:1648]: > sure that's small sample, is it always really small like this?

```
In [2]: N = 10**6

In [3]: %timeit list(np.arange(N))
79.6 ms ± 3.4 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit list(range(N))
32.9 ms ± 1.85 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

```

Comment (jbrockmendel) [reply pandas/core/indexing.py:1648]: wait, better idea, never mind.

Comment (jbrockmendel) [reply pandas/core/indexing.py]:
```
N  = 10

In [5]: %timeit arr = np.asarray(list(range(N)) + [1], dtype=np.intp)
2.52 µs ± 110 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit arr = np.arange(N + 1); arr[-1] = -1
533 ns ± 11 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

N = 10**6

In [8]: %timeit arr = np.arange(N + 1); arr[-1] = -1
397 µs ± 24 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [9]: %timeit arr = np.asarray(list(range(N)) + [1], dtype=np.intp)
98.6 ms ± 437 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Review by jreback — APPROVED** [review]

**Pull request merged**